pixy/dispatcher.go

```diff
@@ -0,0 +1,205 @@
+// requests resolving to it will come in the nearest future.
+type dispatcher struct {
+	contextID         *sarama.ContextID
+	config            *Config
+	factory           dispatchTierFactory
+	requestsCh        chan consumeRequest
+	children          map[string]*expiringDispatchTier
+	expiredChildrenCh chan dispatchTier
+	stoppedChildrenCh chan dispatchTier
+	wg                sync.WaitGroup
+}
+
+// dispatchTier defines a consume request handling tier interface.
+type dispatchTier interface {
+	// key returns the dispatch key of the tier.
+	key() string
+	// requests returns a channel to send requests dispatched to the tier to.
+	requests() chan<- consumeRequest
+	// start spins up the tier's goroutine(s).
+	start(stoppedCh chan<- dispatchTier)
+	// stop makes all tier goroutines stop and releases all resources.
+	stop()
+}
+
+// dispatchTierFactory defines an interface to create dispatchTiers.
+type dispatchTierFactory interface {
+	// dispatchKey returns a key that a child created for the specified `req`
+	// should have.
+	dispatchKey(req consumeRequest) string
+	// newDispatchTier creates a new dispatch tier to handle requests with the
+	// specified dispatch key.
+	newDispatchTier(key string) dispatchTier
+}
+
+// expiringDispatchTier represents a dispatch tier that expires if not used.
+// If a request comes when the tier has already expired but has not yet stopped,
+// then a successor instance is created and new requests are queued to it.
+// as soon as the original tier is stopped the successor is started to take its
+// place.
+type expiringDispatchTier struct {
+	d         *dispatcher
+	factory   dispatchTierFactory
+	instance  dispatchTier
+	successor dispatchTier
+	timer     *time.Timer
+	expired   bool
+}
+
+func newDispatcher(baseCID *sarama.ContextID, factory dispatchTierFactory, config *Config) *dispatcher {
+	d := &dispatcher{
+		contextID:         baseCID.NewChild("dispatcher"),
+		config:            config,
+		factory:           factory,
+		requestsCh:        make(chan consumeRequest, config.ChannelBufferSize),
+		children:          make(map[string]*expiringDispatchTier),
+		expiredChildrenCh: make(chan dispatchTier, config.ChannelBufferSize),
+		stoppedChildrenCh: make(chan dispatchTier, config.ChannelBufferSize),
+	}
+	return d
+}
+
+func (d *dispatcher) start() {
+	spawn(&d.wg, d.run)
+}
+
+func (d *dispatcher) stop() {
+	close(d.requestsCh)
+	d.wg.Wait()
+}
+
+func (d *dispatcher) requests() chan<- consumeRequest {
+	return d.requestsCh
+}
+
+// run receives consume requests from the `requests()` channel and dispatches
+// them to downstream tiers based on request dispatch key.
+func (d *dispatcher) run() {
+	defer d.contextID.LogScope()()
+	for {
+		select {
+		case req, ok := <-d.requestsCh:
+			if !ok {
+				goto done
+			}
+			dt := d.resolveTier(req)
+			// Forward the request to the destination dispatch tier, but drop
+			// it if the tier request channel buffer is full.
```

Comment:
Could you add to this comment the reason for dropping it rather than blocking until the channel has room?